### DISCUSSION TITLE
How to get an imaginary or complex mesh discretization

### DISCUSSION THREAD

**Discussion by salaudeen-ya** [idaholab/moose#24392] [Q&A General] [answered]
Hello,

I am working on a project that needs to stretch the mesh discretization in the complex plane.
Any idea on how to get an imaginary/complex values length?

**Comment by GiudGiud** [comment]
Hello

Is this a 3D mesh?

Guillaume

**Comment by salaudeen-ya** [comment]
I actually need it for all dimensions (1D, 2D and 3D)

On Tue, May 16, 2023 at 1:05 PM Guillaume Giudicelli <
***@***.***> wrote:

> Hello
>
> Is this a 3D mesh?
>
> Guillaume
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/24392#discussioncomment-5918949>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJCC7QYDRIH2XSZ3YRDMUSLXGOXXBANCNFSM6AAAAAAYD5U7GM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Yusuf

**Reply by GiudGiud** [reply]
I dont think we can do that. @roystgnr will know for sure.

So you need complex valued mesh positions, not just complex valued variables?

**Reply by salaudeen-ya** [reply]
I need complex values mesh lengths, in particular. But I will be needing complex value variables too.

**Reply by GiudGiud** [reply]
Complex value variables you could just duplicate the variables, both being Real, and be mindful of it with the math. 
For mesh lengths, we could pull the same trick, but it s not as easy as the multi-mesh capability in MOOSE is hardly used (unless you'd be fine with a MultiApp setup, all Real in one app, all imaginary in the other one, this would mean using fixed point iterations to converge equations)

**Reply by lindsayad** [reply]
Yea at its core `Node` is a `Point` which is a `TypeVector<Real>` so there is definitely not support out of the box. You could play around with modifying libMesh, modifying `Point` to be a `TypeVector<Number>` and configuring libmesh to enable complex numbers, but I don't know what might break if you try that

**Reply by salaudeen-ya** [reply]
I see. I will take a look at modifying the libmesh and hopefully, nothing breaks!

**Reply by roystgnr** [reply]
Tons of things will break.

You might have better luck modifying `Real` to be `std::complex<double>`.  I'd still expect things to break, but at least it's imaginable that the breakage won't be too bad to fix.

**Reply by lindsayad** [reply]
In my vague memory there may not be a lot of breakages but the ones that occur will be somewhat annoying such as interfaces to lapack and blas 🤔

**Reply by roystgnr** [reply]
I think all our BLAS stuff is inside `#ifdef LIBMESH_USE_REAL_NUMBERS` - if you're running complex anyway then the code ought to fall back on our simple hand-coded routines.

**Reply by lindsayad** [reply]
oh nice

**Reply by salaudeen-ya** [reply]
Hello,

I have a quick thought about getting complex lengths/planes. I was wondering if it could be possible to have a MaterialProperty specified that can be used to achieve the complex stretching. With this material property to be used in the Kernels and possibly in the mesh, maybe it could be used to distinguish between the real and the imaginary part. A kernel with the customized MaterialProperty would be used in computing the imaginary part. And eventually, the real and the imaginary part would be coupled in the end (which I don't know how, yet). This is just a conception, and I want to know if this could be possible.

Thanks.

**Reply by lindsayad** [reply]
@cticenhour does a lot of complex number splitting into two parts in the electromagnetics module. It seems tenable to me

**Reply by cticenhour** [reply, marked as answer]
Yes, I do things like this regularly in the EM module. A material example of this is the [WaveEquationCoefficient](https://mooseframework.inl.gov/source/materials/WaveEquationCoefficient.html) object, where it provides real and imaginary components of the material value to all objects that require them.

**Comment by salaudeen-ya** [comment]
I get the part of splitting the real and the imaginary part of
the variables.

But for the mesh, I hope it could be done even with the MOOSE mesh
generators

On Tue, May 16, 2023 at 1:22 PM Guillaume Giudicelli <
***@***.***> wrote:

> Complex value variables you could just duplicate the variables, both being
> Real, and be mindful of it with the math.
> For mesh lengths, we could pull the same trick, but it s not as easy as
> the multi-mesh capability in MOOSE is hardly used.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/24392#discussioncomment-5919127>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJCC7Q6GUDD36DJ25P5OT5LXGOZVXANCNFSM6AAAAAAYD5U7GM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Yusuf

**Reply by lindsayad** [reply]
Please respond in thread if you can. It keeps the discussion much cleaner and more linear

**Reply by GiudGiud** [reply]
It's because they're emailing back. GH doesnt sort emails back into threads

**Reply by lindsayad** [reply]
I know. So better to not email
